cli/flox/src/commands/environment.rs

```diff
@@ -1362,20 +1368,79 @@ impl Pull {
-        let pointer_path = dot_flox_path.join(ENVIRONMENT_POINTER_FILENAME);
 
-        fs::create_dir_all(&dot_flox_path).context("Could not create .flox/ directory")?;
+        let dot_flox_parent = dot_flox_path
+            .parent()
+            .context("Could not get .flox/ parent")?;
+        fs::create_dir_all(dot_flox_parent).context("Could not create .flox/ parent directory")?;
+
+        let temp_dot_flox_dir = tempfile::TempDir::with_prefix_in(DOT_FLOX, dot_flox_parent)
+            .context("Could not create temporary directory for cloning environment")?;
+
+        let pointer_path = temp_dot_flox_dir.path().join(ENVIRONMENT_POINTER_FILENAME);
         fs::write(pointer_path, pointer_content).context("Could not write pointer")?;
 
-        let result =
-            ManagedEnvironment::open(flox, pointer, &dot_flox_path).map_err(Self::convert_error);
+        let mut env = {
+            let result = Dialog {
+                message,
+                help_message: None,
+                typed: Spinner::new(|| ManagedEnvironment::open(flox, pointer, &temp_dot_flox_dir)),
+            }
+            .spin()
+            .map_err(Self::convert_error);
+
+            match result {
+                Err(err) => {
+                    fs::remove_dir_all(&temp_dot_flox_dir)
+                        .context("Could not clean up .flox/ directory")?;
+                    Err(err)?
+                },
+                Ok(env) => env,
+            }
+        };
+
+        let result = Dialog {
+            message,
+            help_message: None,
+            typed: Spinner::new(|| env.build(flox)),
+        }
+        .spin();
+
         match result {
-            Err(err) => {
-                fs::remove_dir_all(dot_flox_path).context("Could not clean up .flox/ directory")?;
-                Err(err)?;
+            Ok(_) => {
+                fs::rename(temp_dot_flox_dir, dot_flox_path)
+                    .context("Could not move .flox/ directory")?;
+            },
+            Err(
+                e @ EnvironmentError2::Core(CoreEnvironmentError::LockedManifest(
+                    LockedManifestError::BuildEnv(CallPkgDbError::PkgDbError(PkgDbError {
+                        exit_code: 100,
+                        ..
+                    })),
+                )),
+            ) => {
+                if !add_systems && !query_add_system(&flox.system) {
+                    bail!(formatdoc! {"
+                        This environment is not yet compatible with your system ({system}).
+
+                        {err:#}
+
+                        Use 'flox pull --add-system' to add your system to the manifest."
+                    , system = flox.system, err = anyhow!(e)});
+                }
+                let doc = amend_current_system(&env, flox)?;
+                env.edit_unsafe(flox, doc.to_string())?;
+
+                fs::rename(temp_dot_flox_dir, dot_flox_path)
```

Comment:
suggestion blocking: I don't think this sets up all the branches and gcroots correctly. They'll still be pointing at the temp `.flox`. No need to fix that now, but at least add a comment

Reply:
the gcroots are not being set at all currently.
We can set a _correct_ gcroot if the build of the modified env succeeds though...

Reply:
Maybe not gcroots, but `ManagedEnvironment::open(flox, pointer, &temp_dot_flox_dir)` will call `ensure_branch` and `ensure_reverse_link` for the wrong path

Reply:
oh you mean `encode` is wrong because it will put the files into the temp folder's path?

Reply:
Yeah, we'll call `encode` with the temp path not the renamed one.

Reply:
though it will adjust itself once you `open` the environment again... 🤔
If this is particularly weird, we can pull into `.flox` directly without using the temp at all there
I just wanted to avoid that we have have the `user/repo.encode(path)` branch populated if the environment is broken and discarded

Reply:
Yeah most of the time it will be fine, so I don't care if you leave it as is, but just leave a comment if you do

Reply:
added comment above where we create the tempdir